contents/javascript/Javascript-framework-React.md

```diff
@@ -0,0 +1,189 @@
+        </div>
+      );
+    };
+    ```
+
+As you can see, React components are just JavaScript functions that accept arbitrary input and return a declarative view describing what should appear in the application's UI. 
+
+#### One-Way Data Binding
+
+*Data binding* refers to the exchange of information between the application data and the UI. React provides *one-way* data binding. In applications that use one-way data binding, changes to the application data are automatically reflected in the UI. However, changes to the UI need to be manually propagated to the application data. 
+
+In the above application, you can see that the UI is updated whenever a different name is entered in the textbox. These updates do not happen in a single step - the application data is first updated using the `handleChange()` method, and then the UI is updated to reflect these changes. 
+
+### Why use React?
+
+Now that we know what React is, let us take a look at some of its benefits.
+
+#### Benefit 1: Abstraction
+
+The traditional <tooltip content="Imperative programming focuses on explicitly describing how a program operates.">*imperative*</tooltip> approach of building a web application requires you to describe how you want your UI to change when your application data changes. 
+
+Most web applications usually have to interact with a lot of <tooltip content="The Document Object Model, or the 'DOM', is an interface to web pages. It represents the page as nodes and objects, allowing programs to read and manipulate the page's content, structure, and styles.">DOM</tooltip> elements and events. Also, different browsers have variations in their implementations of the [core DOM](https://quirksmode.org/dom/core/), [events system](https://www.quirksmode.org/dom/events/index.html) and even [styles](https://www.quirksmode.org/dom/w3c_css.html)! If you want your application to work across different browsers, you would need to manually take care of these variations with the imperative approach. 
+
+React's declarative approach simplifies this process because it abstracts the complexity of interacting with the actual DOM elements and events. For example, it provides its own [event system](https://reactjs.org/docs/events.html) so that events can work in the same way across different browsers.
+
+#### Benefit 2: Better Performance
+
+Web applications can have a lot of user interaction and data updates, which results in changes being made to the DOM. Adding and removing DOM nodes isn't slow, but the performance-bottleneck arises because the browser needs to *[reflow and repaint](https://www.phpied.com/rendering-repaint-reflowrelayout-restyle/)* the UI every time the DOM is changed.
+
+React minimizes this update time by using a *virtual DOM*. The virtual DOM is a JavaScript object that is kept in the memory of your application.
+
+![React's Virtual DOM](javascript-framework-react-imgs/ReactVirtualDOM.png "React's Virtual DOM")
+_Figure 1. How React's actual DOM gets updated._
```

Comment:
The diagram is a bit hard to follow. Show actual and virtual DOMs slightly differently?